l1-contracts/contracts/bridgehub/MessageRootBase.sol

```diff
@@ -112,25 +135,111 @@ abstract contract MessageRootBase is IMessageRoot, Initializable {
         _;
     }
 
-    /// @notice Adds a new chainBatchRoot to the chainTree.
-    /// @param _chainId The ID of the chain whose chainBatchRoot is being added to the chainTree.
-    /// @param _batchNumber The number of the batch to which _chainBatchRoot belongs.
-    /// @param _chainBatchRoot The value of chainBatchRoot which is being added.
-    function addChainBatchRoot(uint256 _chainId, uint256 _batchNumber, bytes32 _chainBatchRoot) external virtual {}
+    function _initialize() internal {
+        // slither-disable-next-line unused-return
+        sharedTree.setup(SHARED_ROOT_TREE_EMPTY_HASH);
+        _addNewChain(block.chainid, 0);
+    }
+
+    function _v30InitializeInner(uint256[] memory _allZKChains) internal {
+        uint256 allZKChainsLength = _allZKChains.length;
+        for (uint256 i = 0; i < allZKChainsLength; ++i) {
+            uint256 batchNumberToWrite = V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY;
+            if (IBridgehubBase(_bridgehub()).settlementLayer(_allZKChains[i]) == L1_CHAIN_ID()) {
+                /// If we are settling on L1.
+                batchNumberToWrite = V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_L1;
+            }
+            v30UpgradeChainBatchNumber[_allZKChains[i]] = batchNumberToWrite;
+        }
+    }
+
+    function saveV30UpgradeChainBatchNumber(uint256 _chainId) external onlyChain(_chainId) {
+        require(block.chainid == IBridgehubBase(_bridgehub()).settlementLayer(_chainId), OnlyOnSettlementLayer());
+        uint256 totalBatchesExecuted = IGetters(msg.sender).getTotalBatchesExecuted();
+        require(totalBatchesExecuted > 0, TotalBatchesExecutedZero());
+        require(
+            totalBatchesExecuted != V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY &&
+                totalBatchesExecuted != V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_L1,
+            TotalBatchesExecutedLessThanV30UpgradeChainBatchNumber()
+        );
+        require(
+            v30UpgradeChainBatchNumber[_chainId] == V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY ||
+                v30UpgradeChainBatchNumber[_chainId] == V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_L1,
+            V30UpgradeChainBatchNumberAlreadySet()
+        );
+        require(currentChainBatchNumber[_chainId] == 0, CurrentBatchNumberAlreadySet());
+
+        currentChainBatchNumber[_chainId] = totalBatchesExecuted;
```

Comment:
maybe here and in similar places it makes sense to also populate `chainBatchRoots` for the current batch? Because usually we would store the root for the latest batch

Reply:
similar places, do you mean chain migrations in setMigratingChainBatchRoot? I think we only save the chainBatchRoots on the current SL and only after V30, so I don't think I would do it